src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -277,22 +292,14 @@ public override bool TryRead(out ReadResult result)
         {
             ThrowIfCompleted();
 
-            return TryReadInternal(InternalTokenSource, out result);
+            return TryReadInternal(out result);
         }
 
-        private bool TryReadInternal(CancellationTokenSource source, out ReadResult result)
+        private bool TryReadInternal(out ReadResult result)
         {
-            bool isCancellationRequested = source.IsCancellationRequested;
-            if (isCancellationRequested || _bufferedBytes > 0 && (!_examinedEverything || _isStreamCompleted))
+            if (_awaitable.IsCompleted || (_bufferedBytes > 0 && (_awaitable.IsCompleted || _isStreamCompleted)))
             {
-                if (isCancellationRequested)
-                {
-                    ClearCancellationToken();
-                }
-
-                ReadOnlySequence<byte> buffer = _readHead == null ? default : GetCurrentReadOnlySequence();
-
-                result = new ReadResult(buffer, isCancellationRequested, _isStreamCompleted);
+                result = GetReadResult();
```

Comment:
So how is StreamPipeReader.TryRead() supposed to work? If no one ever calls StreamPipeReader.ReadAsync() it just forever no-ops and returns false, right? That doesn't seem good.

And if you have to call StreamPipeReader.ReadAsync() anyway, why not just await the result? If you try not to await the result StreamPipeReader.ReadAsync() and then call StreamPipeReader.TryRead() until it returns true, not only is that inefficient, it's likely to be racy since StreamPipeReader doesn't lock when reading or updating fields like _readHead/Tail.

Reply:
TryRead is when you may have not consumed all of the last read? 

e.g. You consume the first of 16 requests, or 4kB file read up to first delimiter, line terminator etc?

Reply:
I'm pretty sure TryRead is supposed to be a non-blocking, polling-based read API. You could use it to consume the rest of the last read, but that's a very limited subset of its intended functionality.

Reply:
The behavior is mirroring what exists in DefaultPipeReader. It will return false if the everything was examined. This PR is not changing that behavior so focus on the cancellation changes and file other issues for behavior changes outside the scope of this PR

Reply:
This absolutely does not mirror the behavior that exists in DefaultPipeReader.

With DefaultPipeReader, you could *only* call TryRead and still get data. That's not the case with StreamPipeReader today or after this PR.

Reply:
File an issue, it should definitely be resolved but not in this PR

Reply:
#39239

The addition of the PipeAwaitable field makes this much easier to implement properly now. If you're not going to fix it as part of the PR, at least through a NotSupportedException in the interim. The current implementation is way too fragile to even think about shipping.

Reply:
Nope. Will fix in a later change. It been this way for a couple of previews and then behavior isn’t damaging it just returns false more often than it needs to.

Reply:
> it just returns false more often than it needs to

And races with the completion of InnerStream.ReadAsync if the caller ever called CancelPendingRead which is NEW with this change. This is the worst failure mode as well since it leads to corrupted results.

And let's not pretend there were a bunch of people calling StreamPipeReader.TryRead before preview7. There are very few uses of TryRead in general (we know because we would have noticed how broken it was otherwise), and hardly any users of StreamPipeReader at all prior to using it for the request body in preview7 which has yet to ship.

Not to mention that anyone expecting to actually receive a request body and wants to use the new Request.BodyReader.TryRead() API will now loop infinitely/timeout.

I suggest you save yourself a lot trouble down the road and fix it or at least throw a NotSupportedException for now.

Reply:
I’m not changing it in this Pr

Reply:
Why TryRead returns false, the polling should stop. The usage pattern is to drain buffered data until it returns false, then switch to ReadAsync. The fact that there’s no parallel data ingestion is just a side effect of the implementation.

Reply:
> It’s completely unrelated

It's *extremely* related.

Before this PR, any code that was consuming the PipeReader in a "valid" manner, could never call TryRead during a call to StreamPipeReader's internal unsynchronized ReadStreamAsync method that updates _readTail. Now that same valid code could get a corrupted result from StreamPipeReader.TryRead() if they ever call the NEW StreamPipeReader.CancelPendingRead() implementation.

Reply:
I’ll make sure the behavior before this PR is preserved as is without throwing NotSupportedException

Reply:
> The usage pattern is to drain buffered data until it returns false, then switch to ReadAsync.

That's one usage pattern. Calling TryRead on an interval is another.

Reply:
I don’t think TryRead should kick off an async read anyways so I’m not even sure how to avoid it

Reply:
It should kick off an async read.

It's not too different conceptually from calling ReadAsync() and then calling CancelPendingRead() immediately after when read doesn't return a completed task. If that's considered a valid usage of the API, I don't see what could possibly be wrong with having TryRead kick off an async read if none is already currently ongoing and there is no data available.

Reply:
I’m don’t agree.

Reply:
TryRead() giving you data eventually if it's available is definitely a part of any reasonable contract for TryRead(). Having to call other APIs as well is not reasonable.

If someone wanted to call ReadAsync(), they would only call ReadAsync(). Our use case of TryRead() as the slightly faster version of ReadAsync() when we know there is more data available is the really unusual/strange/weird use case.

Reply:
Right, and if data is available, it'll return true. Otherwise, returning false is extremely valid. It doesn't need to kick off any operation and just because you claim it's part of *any reasonable contract* doesn't make it so.

Reply:
Because I disagree with the premise, it's very much like IOCP. There's an API to wait for completion and there's an API to poll for completion. The polling API doesn't need to kick off the operation, some other operation can do that, now I understand that when there's no proactive operation going on it's not as useful but I don't agree that it invalidates the contract. 

It's why your issue titled "Make StreamPipeReader.TryRead usable" is well intentioned but also a little trolly.

Reply:
> Right, and if data is available, it'll return true.

Except that's not the case. The client could be uploading all the request data it wants, and I'd never see it if I only call TryRead().

This is a regression caused by the introduction of StreamPipeReader, since before the StreamPipeReader there'd always be a reader calling InnerStream.ReadAsync() if the app was actually consuming the body whether or not it was through TryRead() or ReadAsync().

Reply:
> Except that's not the case. The client could be uploading all the request data it wants, and I'd never see it if I only called TryRead().

That doesn't matter because there's no proactive read going on. So as far as the pipe is concerned, there is no data.

> This is a regression caused by the introduction of StreamPipeReader, since before the StreamPipeReader there'd always be a reader calling InnerStream.ReadAsync() if the app was actually consuming the body whether or not it was through TryRead() or ReadAsync().

 It's not a regression, it's a different implementation.

Reply:
> It's not a regression, it's a different implementation.

With the risk of a consumer that worked fine in one implementation looping infinitely with the other.

If we're just going to decide to have two different implementations, there should be a *very* good reason for it.

The crazy thing is you already did all the real work to match the existing DefaultPipeReader.TryRead() implementation by implementing CancelPendingRead() the way you did.

Reply:
This PR is still a draft, I'll still contemplating changing Kestrel. There are always going to be differences in the implementations. It's just unfortunate that Kestrel depended on so many implementation details but that's what happens when you only have a single implementation.